feat(cypress): Restructure and add Account Portal section (#1679)
clerk/clerk-docs: docs/testing/cypress/overview.mdx

---
title: Testing with Cypress
description: Use Cypress to write end-to-end tests with Clerk.
---

> [!WARNING]
> Our `@clerk/testing` package only supports end-to-end testing. Unit tests are not supported.

Testing with Cypress requires the use of Clerk's [Testing Tokens](/docs/testing/overview#testing-tokens) to bypass various bot detection mechanisms that typically safeguard Clerk applications against malicious bots. Without Testing Tokens, you may encounter "Bot traffic detected" errors in your requests.

This guide aims to help you set up your environment for creating Clerk-authenticated tests with Cypress.

> [!IMPORTANT]
> Check out the [demo repo](https://github.com/clerk/clerk-cypress-nextjs) that tests a Clerk-powered application using [Testing Tokens](/docs/testing/overview#testing-tokens).

<Steps>
  ### Install `@clerk/testing`

  Clerk's testing package provides integration helpers for popular testing frameworks. Install it by running the following command:

  <CodeBlockTabs options={['npm', 'yarn', 'pnpm']}>
    ```sh {{ filename: 'terminal' }}

>     npm i @clerk/testing --save-dev

Reviewer: ```suggestion
    npm install @clerk/testing --save-dev
```

    ```

    ```sh {{ filename: 'terminal' }}
    yarn add -D @clerk/testing
    ```

    ```sh {{ filename: 'terminal' }}
    pnpm add @clerk/testing -D
    ```
  </CodeBlockTabs>

  ### Set your API keys

  In your test runner, set your publishable and secret key as the `CLERK_PUBLISHABLE_KEY` and `CLERK_SECRET_KEY` environment variables, respectively.

  To find your keys:

>   1. Navigate to the [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).
>   1. In the navigation sidebar, select **API Keys**.

Reviewer: ```suggestion
  1. In the Clerk Dashboard, navigate to the [**API Keys**(https://dashboard.clerk.com/last-active?path=api-keys) page.
```

  1. In the **Quick Copy** section, copy your Clerk publishable and secret key.

  > [!WARNING]
  > Ensure you provide the secret key in a secure manner, to avoid leaking it to third parties. For example, if you are using GitHub Actions, refer to [_Using secrets in GitHub Actions_](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).

  ### Global setup

  To set up Clerk with Cypress, you must call the `clerkSetup()` function in your [`cypress.config.ts`](https://docs.cypress.io/guides/references/configuration) file.

  ```tsx filename="cypress.config.ts"
  import { clerkSetup } from '@clerk/testing/cypress'
  import { defineConfig } from 'cypress'

  export default defineConfig({
    e2e: {
      setupNodeEvents(on, config) {
        return clerkSetup({ config })
      },
      baseUrl: 'http://localhost:3000', // your app's URL
    },
  })
  ```

>   `clerkSetup` will obtain a [Testing Token](/docs/testing/overview#testing-tokens) when your test suite starts, making it available for all subsequent tests to use.

Reviewer: ```suggestion
  `clerkSetup()` will obtain a [Testing Token](/docs/testing/overview#testing-tokens) when your test suite starts, making it available for all subsequent tests to use.
```
